src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/KeyedServiceProviderContainerTests.cs

```diff
@@ -0,0 +1,233 @@
+        }
+
+        [Fact]
+        public void ResolveKeyedServiceSingletonInstance()
+        {
+            var service = new Service();
+            var serviceCollection = new ServiceCollection();
+            serviceCollection.AddKeyedSingleton<IService>("service1", service);
+
+            var provider = CreateServiceProvider(serviceCollection);
+
+            Assert.Null(provider.GetService<IService>());
+            Assert.Same(service, provider.GetKeyedService<IService>("service1"));
+        }
+
+        [Fact]
+        public void ResolveKeyedServiceSingletonInstanceWithKeyInjection()
+        {
+            var serviceKey = "this-is-my-service";
+            var serviceCollection = new ServiceCollection();
+            serviceCollection.AddKeyedSingleton<IService, Service>(serviceKey);
+
+            var provider = CreateServiceProvider(serviceCollection);
+
+            Assert.Null(provider.GetService<IService>());
+            var svc = provider.GetKeyedService<IService>(serviceKey);
+            Assert.NotNull(svc);
+            Assert.Equal(serviceKey, svc.ToString());
+        }
+
+        [Fact]
+        public void ResolveKeyedServiceSingletonInstanceWithAnyKey()
+        {
+            var serviceCollection = new ServiceCollection();
+            serviceCollection.AddKeyedSingleton<IService, Service>(KeyedService.AnyKey);
+
+            var provider = CreateServiceProvider(serviceCollection);
+
+            Assert.Null(provider.GetService<IService>());
+
+            var serviceKey1 = "some-key";
+            var svc1 = provider.GetKeyedService<IService>(serviceKey1);
+            Assert.NotNull(svc1);
+            Assert.Equal(serviceKey1, svc1.ToString());
+
+            var serviceKey2 = "some-other-key";
+            var svc2 = provider.GetKeyedService<IService>(serviceKey2);
+            Assert.NotNull(svc2);
+            Assert.Equal(serviceKey2, svc2.ToString());
+        }
+
+        [Fact]
+        public void ResolveKeyedServiceSingletonInstanceWithKeyedParameter()
+        {
+            var serviceCollection = new ServiceCollection();
+            serviceCollection.AddKeyedSingleton<IService, Service>("service1");
+            serviceCollection.AddKeyedSingleton<IService, Service>("service2");
+            serviceCollection.AddSingleton<OtherService>();
+
+            var provider = CreateServiceProvider(serviceCollection);
+
+            Assert.Null(provider.GetService<IService>());
+            var svc = provider.GetService<OtherService>();
+            Assert.NotNull(svc);
+            Assert.Equal("service1", svc.Service1.ToString());
+            Assert.Equal("service2", svc.Service2.ToString());
+        }
+
+        [Fact]
+        public void ResolveKeyedServiceSingletonFactory()
+        {
+            var service = new Service();
+            var serviceCollection = new ServiceCollection();
+            serviceCollection.AddKeyedSingleton<IService>("service1", (sp, key) => service);
```

Comment:
What would clash exactly? Would there be an ambiguous overload to `AddKeyedSingleton`? I think requiring the "key" parameter is fine.

We don't provide a zero-parameter overload for normal factories, so I think it's okay to have some parameters that are often not used. If it's never used because it can be captured from the outer context, it might be worth removing though.

Reply:
For example, `AddSingleton(this IServiceCollection services, Type serviceType, object implementationInstance)` and `AddKeyedSingleton(this IServiceCollection services, Type serviceType, object key` would clash without the "Keyed" in the name.